### PR TITLE
Feature/robot shape

### PR DESCRIPTION
**Pull request by rolfvdhulst**
I created this class a while ago on my own repository. I tried to move it here without causing too many issues and by keeping behavior consistent, as it was really useful on my branch for a lot of calculations needed in roboteam_world and probably also for roboteam_ai. It's become quite large because I had to port some other class changes too, but I think this is really useful class so it should be worth it.

### CODE REVIEW

File: test/utils/RobotShapeTest.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
+//
+// Created by rolf on 18-10-20.
+//
+
+#include <roboteam_utils/RobotShape.h>
+#include <gtest/gtest.h>
+
+#include <roboteam_utils/LineSegment.h>
+using namespace rtt;
```
Comment (mamiksik) [inline test/utils/RobotShapeTest.cpp:10]: new line?

File: test/utils/RobotShapeTest.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+TEST(RobotShapeTest,constructors){
+    RobotShape shape(Vector2(0,0),0.05,0.1,Angle(0.0));
+    EXPECT_EQ(shape.pos(),Vector2(0,0));
+    EXPECT_DOUBLE_EQ(shape.angle().getValue(),0.0);
+    EXPECT_EQ(shape.centerOfKickerPos(),Vector2(0.05,0));
+    LineSegment kicker = shape.kicker();
+    double diff = sqrt(0.1*0.1-0.05*0.05);
+
+    //Thesse are very crucial that they are in the right order,
```
Comment (mamiksik) [inline test/utils/RobotShapeTest.cpp]: typo

File: test/utils/RobotShapeTest.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+#include <roboteam_utils/LineSegment.h>
+using namespace rtt;
+TEST(RobotShapeTest,constructors){
+    RobotShape shape(Vector2(0,0),0.05,0.1,Angle(0.0));
+    EXPECT_EQ(shape.pos(),Vector2(0,0));
+    EXPECT_DOUBLE_EQ(shape.angle().getValue(),0.0);
+    EXPECT_EQ(shape.centerOfKickerPos(),Vector2(0.05,0));
+    LineSegment kicker = shape.kicker();
+    double diff = sqrt(0.1*0.1-0.05*0.05);
```
Comment (mamiksik) [inline test/utils/RobotShapeTest.cpp]: What are these numbers?

File: src/utils/RobotShape.cpp [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            //As the circle is convex, this must mean there is an intersection between the two.
+            Vector2 dribblerIntersect = *kickerLine.firstIntersects(
+                    segment);
+            //Check which of the intersections was invalid.
+            if (firstInFront) {
+                return {dribblerIntersect, intersects[1]};
+            }
+            return {intersects[0], dribblerIntersect};
+        }
+        return {};//This should never be hit as circle intersection always hits atleast two.
```
Comment (mamiksik) [inline src/utils/RobotShape.cpp:121]: Log error?

File: src/utils/Polynomial.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        } else if (discriminant == 0) {
+            return {-b / (2 * a)};
+        }
+        discriminant = sqrt(discriminant);
+        double t0 = (-b - discriminant) / (2 * a);
+        double t1 = (-b + discriminant) / (2 * a);
+        return a > 0 ? std::vector{t0, t1} : std::vector{t1, t0};
+    }
+
+    int numberQuadraticSolutions(double a, double b, double c) {
```
Comment (mamiksik) [inline src/utils/Polynomial.cpp]: countQuadraticSolutions?

File: test/utils/StaticVectorTest.cpp [hunk 6]
```diff
@@ -85,6 +85,7 @@ struct MockClass{
     double value;
 };
 static_assert(std::is_default_constructible_v<MockClass>,"Is type default constructable");
+
```
Comment (mamiksik) [inline test/utils/StaticVectorTest.cpp]: Why this line?

**Comment by mamiksik** [comment]
Hey, I wasn't looking to much into codestyle this time. The code quality overall looks good. Tests are very sweet! 

I suggest squishing the commits into one (I think commits like _fix warning_, _add suggested change_ don't carry much value in git history). and adding one descriptive commit message explaning what is this merge traning to achieve, why, what decision have been made etc. (I recomend: https://chris.beams.io/posts/git-commit/. It's an amazing article about commit messages!)

**Comment by rolfvdhulst** [comment]
We can squish the commits when we merge, no? That way that commit gets a specific message. Please give it a specific review so I can merge; it's been open for almost 1.5 years and we need the PR in roboteam_world..

File: test/StateMachineTest.cpp [hunk 7]
```diff
@@ -131,4 +140,32 @@ TEST(StateMachineTest, test_const_for_loop) {
...
+TEST(StateMachineTest, test_skip_to) {
+    rtt::collections::state_machine<Base, TestEnum, SkillInfo> machine{
+            Base(),
+            Derived(),
+    };
+    machine.skip_to(1);
+    EXPECT_EQ(machine.get_current()->type_num(), 1);
+    machine.skip_to(0);
+    EXPECT_EQ(machine.get_current()->type_num(), 0);
+    //test clamping: TODO: these segfault (not clamped correctly?)
```
Comment (alexander-de-ranitz) [inline test/StateMachineTest.cpp]: Is this TODO still relevant? Or was this solved in this fix?

File: include/roboteam_utils/LineSegment.h [hunk 8]
```diff
@@ -122,6 +126,21 @@ class LineSegment {
...
+    [[nodiscard]] Vector2 direction() const;
+
+
+    /**
+      * Moves the line's start and end points by the given vector
+      */
+    void move(const Vector2& by);
+
+
+    [[nodiscard]] std::optional<Vector2> firstIntersects(const LineSegment &segment) const;
```
Comment (alexander-de-ranitz) [inline include/roboteam_utils/LineSegment.h:146]: Documentation would be nice. It is not immediately obvious what this function does.

File: include/roboteam_utils/LineSegment.h [hunk 9]
```diff
@@ -122,6 +126,21 @@ class LineSegment {
...
+
+    [[nodiscard]] bool preciseDoesIntersect(const LineSegment&line) const;
```
Comment (alexander-de-ranitz) [inline include/roboteam_utils/LineSegment.h:153]: Documentation would be nice. What is the difference with the non-precise doesIntersect and when should which be used?

File: test/utils/Vector2Test.cpp [hunk 10]
```diff
@@ -138,20 +138,20 @@ TEST(VectorTests, moreOperators) {
...
+//    EXPECT_DOUBLE_EQ(x.y(), f.y);
+//    Vector2 checkF = x;
+//    Vector2 b = Vector2(1, 1);
+//    b = x;
+//    EXPECT_DOUBLE_EQ(checkF.x, f.x);
+//    EXPECT_DOUBLE_EQ(checkF.y, f.y);
+//    EXPECT_DOUBLE_EQ(b.x, f.x);
+//    EXPECT_DOUBLE_EQ(b.y, f.y);
+//    std::cout << checkF << std::endl;  // testing print functionality
+//}
```
Comment (alexander-de-ranitz) [inline test/utils/Vector2Test.cpp]: Why was this test commented out? If it isn't useful, perhaps remove it.

File: src/utils/Polynomial.cpp [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        double t0 = (-b - discriminant) / (2 * a);
+        double t1 = (-b + discriminant) / (2 * a);
+        return std::make_pair(t0, t1);
+    }
+
+    std::vector<double> solveQuadraticVector(double a, double b, double c) {
+        double discriminant = b * b - 4 * a * c;
+        if (discriminant < 0) {
+            return {};
+        } else if (discriminant == 0) {
```
Comment (alexander-de-ranitz) [inline src/utils/Polynomial.cpp:34]: Why do this check only here? Is this not also useful in solveQuadratic and solveQuadraticPositiveA?

Comment (alexander-de-ranitz) [reply src/utils/Polynomial.cpp:34]: In general, what is the difference between solveQuadratic and solveQuadraticVector? Functionality seems to essentially be the exact same.

**Review by mamiksik — CHANGES_REQUESTED** [review]

File: test/utils/RobotShapeTest.cpp (hunk 3 above)
Comment (rolfvdhulst) [reply test/utils/RobotShapeTest.cpp]: Fixed them to be variables at the start of the test (no more hardcoded numbers)

File: test/utils/StaticVectorTest.cpp (hunk 6 above)
Comment (rolfvdhulst) [reply test/utils/StaticVectorTest.cpp]: idk.

File: test/utils/Vector2Test.cpp (hunk 10 above)
Comment (rolfvdhulst) [reply test/utils/Vector2Test.cpp]: The function was removed on development (I removed the test now).

File: src/utils/Polynomial.cpp (hunk 11 above)
Comment (rolfvdhulst) [reply src/utils/Polynomial.cpp:34]: Functionality is very similar, but vector only returns as many roots as there actually are. Sometimes, a pair may be more convenient and sometimes a vector is more convenient. In this case, the vector might in particular return only one value whereas the make_pair always returns two (this is sometimes useful, as in some cases you always need to check both, and in some cases you only need the highest/lowest etc.).

**Review by mamiksik — APPROVED** [review]

**Pull request merged**